File: Sources/LoafCat/Modules/HydrationModule.swift
import AppKit

/// The same idea as the stretch break, at a hundredth of the volume: the cat bobs
/// up, says something about water, and goes back to what it was doing.
///
/// No window resize on purpose. A drink of water takes four seconds; taking over the
/// middle of the screen for it would be out of proportion, and the whole reason the
/// stretch break can afford to be dramatic is that nothing else is.
final class HydrationModule: CatModule {
    let id = "hydration"

    private let bus: WellnessBus
    private var nextFire: CFAbsoluteTime = 0
    private var bobUntil: CFAbsoluteTime = 0
    private var bobStart: CFAbsoluteTime = 0
    private var nextLine = 0

    private let bobDuration: Double = 1.1
    private let holdSeconds: Double = 5

    /// Copy, not geometry, so it stays in Swift where it can be localised — the
    /// atlas describes the cat's body, not its vocabulary.
    private let lines = [
        "Water break!",
        "Drink something.",
        "Hydrate, human.",
        "Refill your glass?",
    ]

    init(bus: WellnessBus) {
        self.bus = bus
        settingsChanged()
    }

    func settingsChanged() {
        guard let iv = interval else { nextFire = .greatestFiniteMagnitude; return }
        nextFire = CFAbsoluteTimeGetCurrent() + iv
    }

    private var interval: Double? {
        bus.interval(userMinutes: bus.settings.hydrationMinutes,
                     demoSeconds: bus.isDemo ? 30 : 0)
    }

    func update(_ ctx: TickContext) -> ModuleOutput {
        let now = CFAbsoluteTimeGetCurrent()

        if now >= nextFire, let iv = interval {
            if bus.busy {
                // Never interrupt a stretch break; try again shortly.
                nextFire = now + 5
            } else if ctx.secondsSinceKey > bus.awaySeconds {
                nextFire = now + iv
                bus.log(String(format: "hydration SKIPPED, away %.0fs", ctx.secondsSinceKey))
            } else {
                nextFire = now + iv
                let line = lines[nextLine % lines.count]
                nextLine += 1
                bus.bubble?.say(line, for: holdSeconds)
                bus.chime()
                bobStart = now
                bobUntil = now + bobDuration
                bus.log("hydration \"\(line)\"")
            }
        }

        guard now < bobUntil else { return .none }
        var out = ModuleOutput()
        let p = CGFloat((now - bobStart) / bobDuration)
        // Two quick hops, decaying — enough to draw the eye without asking for a
        // priority high enough to interrupt anything.
        let envelope = (1 - p) * (1 - p)
        out.offset.y = -abs(sin(.pi * 2 * p)) * envelope * bus.atlas.wellness.bobHeight
        out.squash = 1 + envelope * 0.04 * cos(.pi * 2 * p)
        return out
    }
}

File: windows/LoafCat/Modules/HydrationModule.cs
using System.Runtime.Versioning;

namespace LoafCat.Modules;

/// The same idea as the stretch break, at a hundredth of the volume: the cat bobs up,
/// says something about water, and goes back to what it was doing.
///
/// No window resize on purpose. A drink of water takes four seconds; taking over the
/// middle of the screen for it would be out of proportion, and the whole reason the
/// stretch break can afford to be dramatic is that nothing else is.
[SupportedOSPlatform("windows")]
public sealed class HydrationModule : ICatModule
{
    public string Id => "hydration";

    private readonly WellnessBus _bus;
    private double _nextFire;
    private double _bobUntil;
    private double _bobStart;
    private int _nextLine;

    private const double BobDuration = 1.1;
    private const double HoldSeconds = 5;

    /// Copy, not geometry, so it stays in code where it can be localised — the atlas
    /// describes the cat's body, not its vocabulary.
    private static readonly string[] Lines =
    [
        "Water break!",
        "Drink something.",
        "Hydrate, human.",
        "Refill your glass?",
    ];

    public HydrationModule(WellnessBus bus)
    {
        _bus = bus;
        SettingsChanged();
    }

    public void SettingsChanged()
    {
        if (Interval is not { } iv) { _nextFire = Clock.Never; return; }
        _nextFire = _bus.FirstFire(14, iv);
    }

    private double? Interval => _bus.Interval(
        _bus.Settings.HydrationMinutes, _bus.IsDemo ? 30 : 0);

    public ModuleOutput Update(in TickContext ctx)
    {
        double now = Clock.Now;

        if (now >= _nextFire && Interval is { } iv)
        {
            if (_bus.Busy)
            {
                // Never interrupt a stretch break; try again shortly.
                _nextFire = now + 5;
            }
            else if (ctx.SecondsSinceKey > _bus.AwaySeconds)
            {
                _nextFire = now + iv;
                _bus.LogDemo($"hydration SKIPPED, away {ctx.SecondsSinceKey:0}s");
            }
            else
            {
                _nextFire = now + iv;
                string line = Lines[_nextLine % Lines.Length];
                _nextLine++;
                _bus.Bubble?.Say(line, HoldSeconds);
                _bus.Chime();
                _bobStart = now;
                _bobUntil = now + BobDuration;
                _bus.LogDemo($"hydration \"{line}\"");
            }
        }

        if (now >= _bobUntil) return ModuleOutput.None;

        var outv = new ModuleOutput();
        double p = (now - _bobStart) / BobDuration;
        // Two quick hops, decaying — enough to draw the eye without asking for a
        // priority high enough to interrupt anything.
        double envelope = (1 - p) * (1 - p);
        outv.Offset.Y = -Math.Abs(Math.Sin(Math.PI * 2 * p)) * envelope
                        * _bus.Atlas.Wellness.BobHeight;
        outv.Squash = 1 + envelope * 0.04 * Math.Cos(Math.PI * 2 * p);
        return outv;
    }
}
